### PR TITLE
No Connection: Fix for http client stale connection

### PR DESCRIPTION
**Pull request by abhilashraju**
http_client is not handling broken connections well. Eg: Connection termination by server due to keep alive timeout.

At present client is not aware of connection termination from server. So whenever next event ready to be sent client will try send/receives data over broken connection. After failed operation the client will try to restart the connection after closing the current connection. During the process the ssl shutdown will hang when it is made over a broken connection.

Solution:
Restart the connection using fresh new socket object rather than reusing object.Thus the ssl shutdown over broken connection can be avoided.

Tested:
Developer test has been done using bmc and hmc located in different network and let the keep-alive timeout occured. The connection has been restablished on arrival or new event and events are published succesfully to hmc

Change-Id: Ie983f57cbe6e688176501013401009ef0e057ac2

### CODE REVIEW

File: http/http_client.hpp [hunk 1]
```diff
@@ -497,14 +498,19 @@ class ConnectionInfo : public std::enable_shared_from_this<ConnectionInfo>
         }
 
         // Let's close the connection and restart from resolve.
-        doClose(true);
+        doCleanup(true);
+    }
+    void restartConnection()
+    {
+        BMCWEB_LOG_ERROR << "restartConnection ";
```
Comment (sunharis) [inline http/http_client.hpp:505]: Please enhance this trace by adding the destination ip and port number. And edit as "INFO: restart connection to <host> : <ip>

File: http/http_client.hpp [hunk 2]
```diff
@@ -604,18 +615,12 @@ class ConnectionInfo : public std::enable_shared_from_this<ConnectionInfo>
...
-        const std::string& destIPIn, uint16_t destPortIn, bool useSSL,
-        unsigned int connIdIn) :
-        subId(idIn),
-        connPolicy(connPolicyIn), host(destIPIn), port(destPortIn),
-        connId(connIdIn), conn(iocIn), timer(iocIn)
+    std::unique_ptr<boost::asio::ip::tcp::socket>
+        makeConnection(boost::asio::io_context& iocIn, bool useSsl)
     {
-        if (useSSL)
+        BMCWEB_LOG_ERROR << "makeConnection ssl" << useSsl;
```
Comment (sunharis) [inline http/http_client.hpp:621]: please remove this

File: http/http_client.hpp [hunk 3]
```diff
@@ -635,10 +640,24 @@ class ConnectionInfo : public std::enable_shared_from_this<ConnectionInfo>
             } */
             boost::asio::ssl::context sslCtx{
                 boost::asio::ssl::context::tlsv13_client};
-            sslConn.emplace(conn, sslCtx);
+            sslConn.emplace(*newconn.get(), sslCtx);
             setCipherSuiteTLSext();
+            BMCWEB_LOG_ERROR << "created  ssl Conn" << sslConn.has_value();
```
Comment (sunharis) [inline http/http_client.hpp:645]: Add a space between the statement and value. Also please add the destination and port in the traces

File: http/http_client.hpp [hunk 4]
```diff
@@ -763,9 +782,10 @@ class ConnectionPool : public std::enable_shared_from_this<ConnectionPool>
                 }
                 else
                 {
-                    BMCWEB_LOG_DEBUG << "Reusing existing connection "
-                                     << commonMsg;
-                    conn->doResolve();
+                    BMCWEB_LOG_ERROR << "Reusing existing connection "
```
Comment (sunharis) [inline http/http_client.hpp:785]: INFO: please. This is to identify this as information log and not an error

File: http/http_client.hpp [hunk 5]
```diff
@@ -763,9 +782,10 @@ class ConnectionPool : public std::enable_shared_from_this<ConnectionPool>
...
+                                     << commonMsg << "with state "
```
Comment (sunharis) [inline http/http_client.hpp:786]: need spaces when you add the values. Else it becomes unreadable at journal traces

**Review by sunharis — CHANGES_REQUESTED** [review]
Please apply the commit message 50/72 rule at downstream as well.

File: http/http_client.hpp [hunk 6]
```diff
@@ -364,7 +365,7 @@ class ConnectionInfo : public std::enable_shared_from_this<ConnectionInfo>
         }
 
         timer.cancel();
-        if (ec && ec != boost::asio::ssl::error::stream_truncated)
```
Comment (gtmills) [inline http/http_client.hpp:367]: why remove the  boost::asio::ssl::error::stream_truncated ?

File: http/http_client.hpp (hunk 4 above)
Comment (gtmills) [reply http/http_client.hpp:785]: How often does this log? We aren't going to be spamming the journal a crazy amount ?

File: http/http_client.hpp [hunk 7]
```diff
@@ -524,13 +530,18 @@ class ConnectionInfo : public std::enable_shared_from_this<ConnectionInfo>
         {
             // Now let's try to resend the data
             state = ConnState::retry;
-            doResolve();
+            restartConnection();
```
Comment (gtmills) [inline http/http_client.hpp:533]: I think a comment here and a few other places would make this code more maintainable

File: http/http_client.hpp [hunk 8]
```diff
@@ -149,9 +149,10 @@ class ConnectionInfo : public std::enable_shared_from_this<ConnectionInfo>
     // Ascync callables
     std::function<void(bool, uint32_t, Response&)> callback;
     crow::async_resolve::Resolver resolver;
-    boost::asio::ip::tcp::socket conn;
+    boost::asio::io_context& ioc;
     std::optional<boost::beast::ssl_stream<boost::asio::ip::tcp::socket&>>
         sslConn;
+    std::unique_ptr<boost::asio::ip::tcp::socket> conn;
```
Comment (gtmills) [inline http/http_client.hpp:155]: Why did you need to move this to a pointer?

**Review by gtmills — COMMENTED** [review]
Neat stuff! 
Can you add a few comments to this code..?

File: http/http_client.hpp (hunk 8 above)
Comment (abhilashraju) [reply http/http_client.hpp:155]: because we are not  planning to reuse the socket.  I don't know what benefit we get in reuse. But there are troubles seen after keep-alive time out happens.

File: http/http_client.hpp [hunk 9]
```diff
@@ -497,14 +498,19 @@ class ConnectionInfo : public std::enable_shared_from_this<ConnectionInfo>
         }
 
         // Let's close the connection and restart from resolve.
-        doClose(true);
+        doCleanup(true);
```
[the remark this answers is not in the document]
Comment (abhilashraju) [reply http/http_client.hpp:501]: yes, we can do it. But giving explicit name similar to close looked readable. Also I felt it is easier to add new code (in case needed in future) before calling the shutdown. I am open to remove this function.

File: http/http_client.hpp (hunk 1 above)
Comment (abhilashraju) [reply http/http_client.hpp:505]: Ack

File: http/http_client.hpp (hunk 7 above)
Comment (abhilashraju) [reply http/http_client.hpp:533]: Ack

File: http/http_client.hpp (hunk 2 above)
Comment (abhilashraju) [reply http/http_client.hpp:621]: Ack

File: http/http_client.hpp (hunk 3 above)
Comment (abhilashraju) [reply http/http_client.hpp:645]: Ack

File: http/http_client.hpp (hunk 5 above)
Comment (abhilashraju) [reply http/http_client.hpp:786]: Ack

File: http/http_client.hpp (hunk 3 above)
Comment (sunharis) [reply http/http_client.hpp:646]: Not done?

**Review by sunharis — APPROVED** [review]
Upstream is being worked via https://jsw.ibm.com/browse/PFEBMC-1644

File: http/http_client.hpp [hunk 10]
```diff
@@ -48,7 +48,6 @@
 #include <memory>
 #include <queue>
 #include <string>
-
```
Comment (gtmills) [inline http/http_client.hpp:51]: be careful with empty spare changes. Oh well, for next time. ;)

File: http/http_client.hpp (hunk 8 above)
Comment (gtmills) [reply http/http_client.hpp:154]: Ack

File: http/http_client.hpp (hunk 7 above)
Comment (gtmills) [reply http/http_client.hpp:535]: I think a few more comments, a more descriptive comment on why would help..

**Comment by gtmills** [comment]
What is the upstream plan here?

**Review by gtmills — APPROVED** [review]
I am good with this but can this be pushed upstream?
